src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,156 @@
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(iOutOfBounds);
+        }
+
+        if (index == 0) {
+            head = new Node<>(null, value, head);
+            if (size == 0) {
+                tail = head;
+            }
+        } else if (index == size) {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+        } else {
+            Node<T> previous = getNode(index - 1);
+            Node<T> newNode = new Node<>(previous, value, previous.next);
+            previous.next = newNode;
+            if (newNode.next != null) {
+                newNode.next.prev = newNode;
+            }
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(iOutOfBounds);
+        }
+
+        Node<T> current;
+
+        if (index < size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current.value;
+    }
+
+    private Node<T> getNode(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(iOutOfBounds);
+        }
+
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNode(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(iOutOfBounds);
+        }
+
+        Node<T> removedNode = getNode(index);
+        updateReferences(removedNode);
+        return removedNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if (object == null) {
+                if (current.value == null) {
+                    updateReferences(current);
+                    return true;
+                }
+            } else if (object.equals(current.value)) {
+                updateReferences(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
+    private void updateReferences(Node<T> node) {
```

Comment:
naming not clearly explains how we are updating references
```suggestion
    private void unlink(Node<T> node) {
```